### PR TITLE
Compatibility with dask 2021.02.0

### LINKED ISSUE #4860: ⚠️ Nightly upstream-dev CI failed ⚠️
[Workflow Run URL](https://github.com/pydata/xarray/actions/runs/556342280)
<details><summary>Python 3.8 Test Summary Info</summary>

```
FAILED xarray/tests/test_dask.py::test_persist_Dataset[<lambda>1] - IndexErro...
```

</details>

### PR DESCRIPTION
**Pull request by crusaderky**
Closes #4860
Reverts #4873 

Restore compatibility with dask 2021.02.0 by avoiding improper assumptions on the implementation details of ``da.Array.__dask_postpersist__()``.

This PR *does not* align xarray to the new dask collection spec (https://github.com/dask/dask/issues/7093), as I just realized that Datasets violate the rule of having all dask keys with the same name if they contain more than one dask variable - and cannot do otherwise. So I have to change the dask collection spec again to accommodate them.

### CODE REVIEW

**Review by keewis — APPROVED** [review]
looks good to me. I don't know a lot about `dask`, though, so someone else might need to look at this.

File: xarray/core/dataset.py [hunk 1]
```diff
@@ -920,20 +920,38 @@ def _dask_postcompute(results, info, *args):
...
+                    # dask.graph_manipulation module.
+                    #
+                    # In case of multiple layers, don't pollute a Variable's
+                    # HighLevelGraph with layers belonging exclusively to other
+                    # Variables. However, we need to prevent partial layers:
+                    # https://github.com/dask/dask/issues/7137
+                    # TODO We're wasting a lot of key-level work. We should write a fast
+                    #      variant of HighLevelGraph.cull() that works at layer level
+                    #      only.
+                    dsk2 = dsk.cull(keys)
```
Comment (dcherian) [inline xarray/core/dataset.py:947]: should we add tests for this code path?

Comment (crusaderky) [reply xarray/core/dataset.py:947]: I will once https://github.com/dask/dask/issues/7203 is out. I can remove the code path for now if you prefer.

Comment (dcherian) [reply xarray/core/dataset.py:947]: maybe that's safer?

Comment (crusaderky) [reply xarray/core/dataset.py]: I removed the HLG and reworked the whole thing

**Review by dcherian — APPROVED** [review]

**Pull request merged**